Initial staleness work

brian-brazil opened the pull request:
This is just enough to get the right staleness logic when a target stops exposing a time series, as that's the easiest case. There's still other code paths to be covered (e.g. rules, failed scrape, target disappears) but I wanted to share this much first.

@fabxc

brian-brazil commented:
Here's a screenshot of a series appearing and disappearing: ![screenshot_2017-04-14_11-43-12](https://cloud.githubusercontent.com/assets/7115638/25041397/a8d2add4-2107-11e7-93ca-825b39f73bec.png) (The forward diagonal is returning no points).

#398

fabxc commented:
Generally speaking, this looks good. Little amount of code so far, which is also good.

brian-brazil commented on pkg/value/value.go:
```diff
@@ -0,0 +1,19 @@
+
+package value
+
+var (
```
Done

brian-brazil commented on promql/engine.go:
```diff
@@ -734,6 +735,9 @@ func (ev *evaluator) vectorSelector(node *VectorSelector) Vector {
 				continue
 			}
 		}
+		if math.Float64bits(v) == value.StaleNaN {
```
Done

brian-brazil commented:
Should we get this much committed, or would you prefer I continue working on this until it's more complete?

fabxc commented:
What other things need to be added? A checklist would help.

brian-brazil commented:
This is just changes between scrapes. There's also failed scrapes, rules/alerts, crash handling (which I think we can get away with also covering targets disappearing/changing) and federation.

juliusv commented on retrieval/scrape.go:
```diff
@@ -434,7 +443,8 @@ func newScrapeLoop(ctx context.Context, sc scraper, app, reportApp func() storag
 		reportAppender: reportApp,
-		cache:          map[string]uint64{},
+		refCache:       map[string]uint64{},
+		lsetCache:      map[uint64]lsetCacheEntry{},
```
It looks like these caches are slow memory leaks if long-lived targets change series a lot. Maybe at least add a TODO comment to periodically purge them or something? Actually maybe the same staleness logic can be reused to purge stale cache entries?

brian-brazil replied:
Yeah, I noticed that. I've it down as future work, the tsdb API is going to change too which will require a rejig here. TODO added.

> Actually maybe the same staleness logic can be reused to purge stale cache entries?

More simply you can keep just the entries from the last scrape.

brian-brazil commented:
Rebased and updated to take explicit timestamps into account (which is to say, no staleness markers are applied for series with explicit timestamps - there's no sane way to make it work).

fabxc commented on retrieval/scrape_test.go:
```diff
@@ -437,6 +437,56 @@ func TestScrapeLoopRun(t *testing.T) {
+		t.Fatalf("Appended samples not as expected. Wanted: %f for first sample Got: %f", appender.result[0], 42)
+	}
+	ingestedNaN := math.Float64bits(appender.result[1].v)
+	if ingestedNaN != value.StaleNaN {
```
use `value.IsStaleNaN()`

brian-brazil replied:
Done

fabxc commented:
Do we want the textproto to expose staleness values so we can propagate staleness markers in federation?

brian-brazil commented:
Yes, that's my current thinking on it. That has knock-on effects in terms of protocol definition, and backward compatibility.

brian-brazil commented:
Though that isn't a hard requirement for this to be useful, it could be added later.

juliusv commented on retrieval/scrape.go:
```diff
@@ -547,13 +568,14 @@ loop:
 			case storage.ErrNotFound:
 				ok = false
-			case errSeriesDropped:
+			case errSeriesDropped, storage.ErrOutOfOrderSample, storage.ErrDuplicateSampleForTimestamp:
```
Can we add back logging for the out-of-order and duplicate-sample cases? I think there's not eveng a metric for it anymore at the moment.

brian-brazil replied:
Done.

juliusv commented on pkg/value/value.go:
```diff
@@ -0,0 +1,27 @@
+
+const (
+	NormalNaN uint64 = 0x7ff8000000000001 // A quiet NaN. This is also math.NaN().
+	StaleNaN  uint64 = 0x7ff4000000000000 // A signalling NaN, starting 01 to allow for expansion.
```
Why "starting 01" if it ends with `00`?

brian-brazil replied:
I don't understand your comment.

It needs to start with 0 to be a signaling nan, and then a 1 for the first type of information. The next one if we need them will be 001, 0001 etc.

juliusv replied:
Oh never mind me, I thought you were referring to the last two hex digits of the overall number here - just looked up how `NaN`s are actually encoded and get it now. I also didn't know that "signaling `NaN`s" are an official concept. TIL.

Maybe to help the uninitiated like me, amend the comment to "A signaling NaN, with the significand starting with 01 to allow for expansion.". I guess a link to https://en.wikipedia.org/wiki/NaN somewhere around here would not hurt either :)

Also, is this part of the article unproblematic for us:

> in NaNs generated by the PA-RISC and MIPS processors, the signaling/quiet bit is zero if the NaN is quiet, and non-zero if the NaN is signaling. Thus, on these processors, the bit represents an 'is_signaling' flag.

Go doesn't yet have a MIPS port, but apparently there's [something experimental in the works](https://golang.org/doc/go1.6#ports). I'm wondering if a quiet NaN as we define it could end up as a signalling NaN there and possibly cause the CPU to throw exceptions when doing ops on it?

brian-brazil replied:
We'll only ever be doing bit operations on the signalling NaNs.

> I'm wondering if a quiet NaN as we define it could end up as a signalling NaN there and possibly cause the CPU to throw exceptions when doing ops on it?

That'd break for all of Go then.

brian-brazil replied:
Done, value also changed for Reasons that may come up latter (don't want to hog the value with the most potential for a parameter by a marker that doesn't need a parameter).

brian-brazil commented:
This is all of target staleness, other than `up` and friends which I'll finish off later in the week.

brian-brazil commented:
Okay, that's all the logic required for targets. Please review.

juliusv commented on retrieval/scrape.go:
```diff
@@ -460,18 +472,22 @@ func (sl *scrapeLoop) run(interval, timeout time.Duration, errc chan<- error) {
 			total, added int
 			start        = time.Now()
-			scrapeCtx, _ = context.WithTimeout(sl.ctx, timeout)
+			scrapeCtx, _ = context.WithTimeout(sl.scrapeCtx, timeout)
```
This made me notice that the fix for context cancellation in https://github.com/prometheus/prometheus/pull/2599/files is still out for review...

juliusv commented on retrieval/scrape.go:
```diff
@@ -517,15 +522,65 @@ func (sl *scrapeLoop) run(interval, timeout time.Duration, errc chan<- error) {
+
+	// Call sl.append again with an empty scrape to trigger stale markers.
+	// If the target has since been recreated and scraped, the
+	// stale markers will be out of order and ignored.
```
This puts a hard requirement on local and remote storages to never support out-of-order samples. That will even be infeasible for the usual remote storages, where the adapter (like our example one) doesn't keep state and the remote end supports arbitrarily timed samples?

juliusv commented on retrieval/scrape.go:
```diff
@@ -482,30 +498,92 @@ func (sl *scrapeLoop) run(interval, timeout time.Duration, errc chan<- error) {
+
+	// Call sl.append again with an empty scrape to trigger stale markers.
+	// If the target has since been recreated and scraped, the
+	// stale markers will be out of order and ignored.
```
This puts a hard requirement on local and remote storages to never support out-of-order samples. That will even be infeasible for the usual remote storages, where the adapter (like our example one) doesn't keep state and the remote end supports arbitrarily timed samples?

brian-brazil replied:
I'm presuming that they'll be filtered out before sending them to remote storage.

juliusv replied:
Ok, I think that's acceptable for now, and it doesn't prevent us from solving this better in the future (with explicit target reappearance tracking or something).

brian-brazil replied:
Even with better reappearance tracking, there's corner cases like a time series moving between targets.

fabxc replied:
I don't fully understand this. Do you want to filter out all staleness markers from LTS? Doesn't that defeat large parts of the purpose. Graphs will still be wrong on LTS data.

brian-brazil replied:
We'll send on the staleness markers to LTS.

We'll only send data on to LTS after the commit, as that's what makes sense even without considering staleness handling.

juliusv replied:
Ah, so when you said,

> I'm presuming that they'll be filtered out before sending them to remote storage.

Did you mean that out-of-order samples should be filtered out in the Prometheus code that sends to the remote storage? If so, how (it doesn't keep enough state)?

Initially I thought you meant that the staleness markers would be filtered out, but that doesn't seem to be the case now.

brian-brazil replied:
> If so, how (it doesn't keep enough state)?

The out of order error from the storage indicates not to send on that sample.

juliusv replied:
I see, so you mean we require the local storage to tell us whether we should send something on to the remote one. What if we disable the local storage though (like for Cortex)? Cortex actually will also reject out-of-order samples, but not every LTS will. Just live with it?

brian-brazil replied:
> I see, so you mean we require the local storage to tell us whether we should send something on to the remote one.

Yes.

> What if we disable the local storage though (like for Cortex)? Cortex actually will also reject out-of-order samples, but not every LTS will. Just live with it?

If a LTS doesn't support it, it'd probably be best for them to drop the staleness markers.

Getting staleness to work correctly on a system like Cortex would be challenging, it's only meant to fully work on the Prometheus that's doing the original scrape/evaluation.

juliusv replied:
Ok, so just to be clear on the rule then: staleness markers should be filtered in the *adapter* for any LTS that does not support rejecting out-of-order samples?

But I still don't understand why staleness wouldn't work correctly for Cortex, which does have rejection of out-of-order samples. It would at least work in most cases, right?

brian-brazil replied:
> Ok, so just to be clear on the rule then: staleness markers should be filtered in the adapter  for any LTS that does not support rejecting out-of-order samples?

That'd be one option for such a system where Prometheus doesn't have local storage.

> But I still don't understand why staleness wouldn't work correctly for Cortex, which does have rejection of out-of-order samples. It would at least work in most cases, right?

It'll work in some cases, not all due to additional races such a setup has.

juliusv commented:
:+1: besides my comment about LTS out-of-order requirements...

brian-brazil commented on retrieval/scrape.go:
```diff
@@ -482,30 +498,92 @@ func (sl *scrapeLoop) run(interval, timeout time.Duration, errc chan<- error) {
+			// The append failed, probably due to a parse error.
+			// Call sl.append again with an empty scrape to trigger stale markers.
+			if _, _, err = sl.append([]byte{}, start); err != nil {
+				log.With("err", err).Error("failure append failed")
```
Done

brian-brazil commented on retrieval/scrape.go:
```diff
@@ -460,18 +472,22 @@ func (sl *scrapeLoop) run(interval, timeout time.Duration, errc chan<- error) {
+			close(sl.stopped)
 			return
+		case <-sl.scrapeCtx.Done():
+			break mainLoop
```
Done

brian-brazil commented on retrieval/scrape.go:
```diff
@@ -568,26 +654,63 @@ loop:
+				}
+			}
+		}
+	}
```
If the storage fails for the stale markers and returns an error we should rollback, so I think this is right.

Code is using gofmt.

fabxc replied:
Ah, right, we are setting an error in the `err == nil` block above.

brian-brazil commented on retrieval/scrape.go:
```diff
@@ -482,30 +498,92 @@ func (sl *scrapeLoop) run(interval, timeout time.Duration, errc chan<- error) {
+			log.With("err", err).Error("append failed")
+			// The append failed, probably due to a parse error.
+			// Call sl.append again with an empty scrape to trigger stale markers.
+			if _, _, err = sl.append([]byte{}, start); err != nil {
```
What way were you thinking of splitting them? Parsing and ingesting are fairly tightly bound as-is.

brian-brazil commented on retrieval/scrape.go:
```diff
@@ -568,26 +654,63 @@ loop:
+				continue
+			case storage.ErrDuplicateSampleForTimestamp:
+				err = nil
+				log.With("timeseries", string(met)).Warn("Duplicate sample for timestamp")
```
Done. I plumbed through a logging with the target labels.

I kept the per-sample logs as Debug logs.

fabxc commented on retrieval/scrape.go:
```diff
@@ -531,9 +613,10 @@ func (s samples) Less(i, j int) bool {
+		app            = sl.appender()
+		p              = textparse.New(b)
+		defTime        = timestamp.FromTime(ts)
+		samplesScraped = map[string]labels.Labels{}
```
This generates garbage on every . Probably fine but we should at least initialize it with the size of the previous one. I'd also rename it to `prevSeries` in the scrapeLoop as it records series not samples.

brian-brazil replied:
Done

fabxc commented:
This needs a rebase.

brian-brazil commented:
Rebased

fabxc commented:
Honestly, I cannot review this PR any further. When switching to the code view my browser remains at the brink of locking up and I need 5+ minutes to add a comment. It's not even that big of a change...

But generally, code changes are alright now I think and on the side of semantic correctness you are more competent anyway. So I think we should merge this to move forward.

👍

juliusv commented:
:+1: Same.

Disabling the Sourcegraph Chrome extension helped the browser sluggishness for me though.

brian-brazil commented:
Thanks for the review, once this is merged we'll see if any issues come up semantically.

I'm seeing no problems reviewing this change in my browser.

fabxc commented:
Ah, that was it. Will remember for next time.
